### PR TITLE
[nvJPEG] Integrate nvJPEG decoder

### PR DESCRIPTION
**Pull request by wang-xinyu**
### PR types(PR类型)

Image decoding optimization with nvJPEG

### Description

- Implement ImageDecoder class, support OpenCV and nvJPEG decode
- Fix the GPU memory reAlloc logic in FDTensor
- Compare the performance of OpenCV and nvJPEG
- PaddleClas add new predict APIs on FDMat and image_name
- Use shared_ptr to avoid deep copy of the FDMat.fd_tensor

### CODE REVIEW

**Comment by wang-xinyu** [comment]
## Test Environment
CPU: Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz
GPU: T4
Model: ResNet50 224x224
Backend: TensorRT FP16
Warmup 100 rounds，tested 1000 rounds and get avg. latency.

## cv::imread vs. nvJPEG read + decode

| Image Shape | Batch Size | OpenCV | nvJPEG | Gain |
|-|-|-|-|-|
| 320x240 | 1 | 1.77018ms | 0.597424ms | 66.25%  |
| 640x480 | 1 | 3.56971ms | 1.07261ms | 69.95% |
| 1920x1080 | 1 | 22.5409ms | 5.23673ms | 76.77% |

## Read, Decode and Predict (OpenCV vs. nvJPEG+ CV-CUDA)

| Image Shape | Batch Size | OpenCV | nvJPEG + CV-CUDA | Gain |
|-|-|-|-|-|
| 320x240 | 1 | 2.67243ms | 1.56112ms | 41.58%  |
| 640x480 | 1 | 5.31156ms | 2.16612ms | 59.22% |
| 1920x1080 | 1 | 26.1565ms | 6.21526ms | 76.24% |
| 320x240 | 4 | 8.92629ms | 4.10219ms | 54.04%  |
| 640x480 | 4 | 19.827ms | 6.44255ms | 67.51% |
| 1920x1080 | 4 | 96.4512ms | 21.2363ms | 77.98% |

**Comment by wang-xinyu** [comment]
Tested accuracy on the CI data(50 images) with CI script https://github.com/felixhjh/fastdeploy_ci/blob/main/scripts_ci/test_resnet50_vd.py.

When using nvJPEG + CV-CUDA, the top1 and top5 accuracy are same as OpenCV.

But the scores are slightly different.

- nvJPEG has diff compared with cv::imread (OpenCV 3.4.16)
- CV-CUDA resize has diff compared with cv::resize (OpenCV 3.4.16)

File: FastDeploy.cmake.in [hunk 1]
```diff
@@ -169,21 +169,25 @@ if(ENABLE_POROS_BACKEND)
 endif()
 
 if(WITH_GPU)
-  if (NOT CUDA_DIRECTORY)
+  if(NOT CUDA_DIRECTORY)
```
Comment (jiangjiajun) [inline FastDeploy.cmake.in:172]: 在这里面不需要添加对ENABLE_NVJPEG的判断吗？

**Review by jiangjiajun — APPROVED** [review]

File: FastDeploy.cmake.in (hunk 1 above)
Comment (wang-xinyu) [reply FastDeploy.cmake.in:172]: 这个ENABLE_NVJPEG没有设置为编译选项，只要WITH_GPU并且不是JETSON，就加上了

**Comment by leiqing1** [comment]
> Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz

需要补充cuda版本、cudnn版本、TensorRT版本

**Pull request merged**